### PR TITLE
automation: only read complete lines before trying to deserialize

### PR DESCRIPTION
**Pull request by tgummerer**
When tailing the event log in automation API we currently have nothing that makes sure we read only complete lines.  This means if the OS happens to flush an incomplete line for whatever reason (or the Go JSON encoder does, which we're using to write these lines), we might read a line that is incompletely written, and thus will fail to JSON decode it.

Since the JSON encoder always writes a newline at the end of each string, we can also make sure that the line we read ends with a newline and otherwise wait for the rest of the line to be written.

The library we use in Go provides a convenient setting for this, while in python and nodejs we need to add some code to do this ourselves.

Fixes https://github.com/pulumi/pulumi/issues/15235
Fixes https://github.com/pulumi/pulumi/issues/15652
Fixes https://github.com/pulumi/pulumi/issues/9269 (This is closed already, but never had a proper resolution afaics)
Fixes https://github.com/pulumi/pulumi/issues/6768

It would be nice to add a typescript test here as well, but I'm not sure how to do that without marking the readLines function non-private.  But I don't know typescript well, so any hints of how to do that would be appreciated!

## Checklist

- [x] I have run `make tidy` to update any new dependencies
- [x] I have run `make lint` to verify my code passes the lint check
  - [x] I have formatted my code using `gofumpt`

- [x] I have added tests that prove my fix is effective or that my feature works  

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

**Review by Frassle — APPROVED** [review]

**Comment by Frassle** [comment]
Guess we need a similar fix in dotnet?

File: sdk/nodejs/automation/stack.ts [hunk 1]
```diff
@@ -123,17 +123,17 @@ export class Stack {
...
+		event = JSON.parse(line);
                 callback(event);
             } catch (e) {
-                log.warn(`Failed to parse engine event
-If you're seeing this warning, please comment on https://github.com/pulumi/pulumi/issues/6768 with the event and any
-details about your environment.
-
-Event: ${line}\n${e.toString()}`);
+                partialLine += line;
+                return;
```
Comment (tgummerer) [inline sdk/nodejs/automation/stack.ts:136]: Unfortunately `nodejs` doesn't seem to include the newline character in the on line events, so we'll have to go off catching exceptions here.

**Comment by tgummerer** [comment]
https://github.com/pulumi/pulumi-dotnet/pull/245 is the corresponding dotnet PR.

**Comment by justinvp** [comment]
> It would be nice to add a typescript test here as well, but I'm not sure how to do that without marking the readLines function non-private. But I don't know typescript well, so any hints of how to do that would be appreciated!

> Unfortunately `nodejs` doesn't seem to include the newline character in the on line events, so we'll have to go off catching exceptions here.

Yeah, we should have a test for Node.js. It doesn't look like `readLines` needs to be a private member of the class -- I think we could move it out to be a top-level exported function marked `/** @internal */` so it can be called from a test.

```typescript
/** @internal */
export async function readLines(logPath: string, callback: (event: EngineEvent) => void): Promise<ReadlineResult> {
    // ...
}
```

**Pull request merged**